### PR TITLE
upgrade polkadot to polkadot-stable2506 and upgrade toolchain version

### LINKED ISSUE #2521: Upgrade to Polkadot Release polkadot-stable2506
The Parity team has published new version of Polkadot: `polkadot-stable2506`

See https://github.com/paritytech/polkadot-sdk/releases/polkadot-stable2506

## Upgrade Priority

See [Polkadot Release Page](https://github.com/paritytech/polkadot-sdk/releases/polkadot-stable2506)

## Host Functions

See [Polkadot Release Page](https://github.com/paritytech/polkadot-sdk/releases/polkadot-stable2506)

## Database Migrations

See [Polkadot Release Page](https://github.com/paritytech/polkadot-sdk/releases/polkadot-stable2506)

## Runtime Migrations

See [Polkadot Release Page](https://github.com/paritytech/polkadot-sdk/releases/polkadot-stable2506)

### PR DESCRIPTION
**Pull request by saraswatpuneet**
# Goal
The goal of this PR is 

Closes #2521
 
# Discussion

# Checklist
- [ ] Updated Pallet Readme?
- [ ] Updated js/api-augment for Custom RPC APIs?
- [ ] Design doc(s) updated?
- [ ] Unit Tests added?
- [ ] e2e Tests added?
- [ ] Benchmarks added?
- [ ] Spec version incremented?

### CODE REVIEW

**Review by JoeCap08055 — COMMENTED** [review]
If we're moving Rust versions, do we also need or want to update the nightly toolchain that we use for formatting and linting? It's specified in a couple places in the CI workflows, as well as the main `Makefile`.

**Review by JoeCap08055 — APPROVED** [review]
🚢 it!

File: runtime/frequency/src/lib.rs [hunk 1]
```diff
@@ -1406,6 +1404,7 @@ impl cumulus_pallet_parachain_system::Config for Runtime {
 	type WeightInfo = ();
 	type ConsensusHook = ConsensusHook;
 	type SelectCore = DefaultCoreSelector<Runtime>;
+	type RelayParentOffset = ConstU32<0>;
```
Comment (aramikm) [inline runtime/frequency/src/lib.rs:1407]: nit: Are we sure we want to set this to zero? It might help with forks @wilwade 

```
		/// While this helps to reduce forks on the parachain side, it increases the delay for
		/// processing XCM messages. So, the value should be chosen wisely.

```

**Review by aramikm — APPROVED** [review]
Overall looks good just a few comments.

- You should probably change the following line in docker-compose.yml
`x-polkadot-image: &polkadot-image "parity/polkadot:stable2503-7"`
- We might also need to rerun the benchmarks on this PR
- We will need to crate a new ticket and re-calculate our DB read and writes Benchmarks from the latests mainnet DB.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -368,7 +368,7 @@ jobs:
           profile: ${{matrix.build-profile}}
           package: ${{matrix.package}}
           chain: ${{matrix.chain}}
-          tag: "1.84.1"
+          tag: "1.88.0"
```
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply .github/workflows/release.yml:371]: There was conflict around srtool using 1.88 now so instead of just updating just that, decided to overhaul the whole base.

File: common/primitives/src/offchain.rs [hunk 3]
```diff
@@ -22,7 +22,7 @@ sp_externalities::decl_extension! {
 #[runtime_interface]
 pub trait Custom: ExternalitiesExt {
 	/// another function
-	fn get_val(&mut self) -> Option<Vec<u8>> {
+	fn get_val(&mut self) -> AllocateAndReturnByCodec<Option<Vec<u8>>> {
```
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply common/primitives/src/offchain.rs:25]: Good catch, I will check it again

File: runtime/frequency/src/lib.rs (hunk 1 above)
Comment (saraswatpuneet) [reply runtime/frequency/src/lib.rs:1407]: Read somewhere that for most cases 0 is okay but yes this is one thing we should investigate, @enddynayn  might have some ideas

File: Cargo.toml [hunk 4]
```diff
@@ -138,52 +138,53 @@ unicode-normalization = { version = "0.1.22", default-features = false }
...
+substrate-frame-rpc-system = { git = "https://github.com/paritytech/polkadot-sdk", tag = "polkadot-stable2506" }
+substrate-prometheus-endpoint = { git = "https://github.com/paritytech/polkadot-sdk", tag = "polkadot-stable2506" }
+sp-debug-derive = { git = "https://github.com/paritytech/polkadot-sdk", tag = "polkadot-stable2506", default-features = false }
 
 # Lints: https://doc.rust-lang.org/rustc/lints/index.html
 # Lint Levels: https://doc.rust-lang.org/rustc/lints/levels.html
 # Clippy Lints: https://rust-lang.github.io/rust-clippy/master/index.html
 [workspace.lints.rust]
 nonstandard-style = "deny"
+unexpected_cfgs = { level = "warn", check-cfg = ['cfg(substrate_runtime)'] }
```
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply Cargo.toml:187]: By default, Clippy  warn about "unexpected" or "unknown" configuration flags that aren't built into the standard library, since this is more of a warning around this custom feature.

File: common/primitives/src/offchain.rs [hunk 5]
```diff
@@ -18,14 +18,30 @@ sp_externalities::decl_extension! {
...
+				let encoded = ext.0.clone().encode();
+				let len = encoded.len().min(256);
+				output.clear();
+				match output.try_extend_from_slice(&encoded[..len]) {
+					Ok(()) => len as u32,
+					Err(_) => 0,
+				}
+			},
+			None => 0,
+		}
```
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply common/primitives/src/offchain.rs:45]: Keeping the legacy function so runtime can match function signature for older node but legacy function will return None in new node so it falls back on newly introduce function called `get_val_buffered`, this way older nodes will still be able to resolve the function without need to enforce upgrades

File: pallets/msa/src/offchain_storage.rs [hunk 6]
```diff
@@ -455,14 +452,30 @@ pub struct FinalizedBlockResponse {
...
-		common_primitives::offchain::custom::get_val()
-			.unwrap_or(RPC_FINALIZED_BLOCK_REQUEST_URL.into())
+		// To prevent breaking change we call legacy function
+		// TODO remove this logic once all collators are upgraded
+		let legacy_val = common_primitives::offchain::custom::get_val();
+		if legacy_val.is_some() {
+			legacy_val.unwrap_or(RPC_FINALIZED_BLOCK_REQUEST_URL.into())
+		} else {
+			// call the runtime-interface function that fills our fixed buffer
+			let mut buffer = vec![0u8; 256];
```
Comment (saraswatpuneet) [inline pallets/msa/src/offchain_storage.rs:466]: send a reasonable sized buffer to offchain and read the url bytes

File: .github/workflows/release.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply .github/workflows/release.yml:371]: looks like i have revert to 1.84.1 subswasm is still behind the latest polkadot

File: common/primitives/src/offchain.rs [hunk 7]
```diff
@@ -22,8 +25,24 @@ sp_externalities::decl_extension! {
...
+		None
+	}
+
+	/// Get extension value by writing to output buffer
+	/// Returns the total length of encoded data,
+	/// or 0 if no extension found.
+	fn get_val_buffered(&mut self, output: PassFatPointerAndReadWrite<&mut [u8]>) -> u32 {
+		match self.extension::<OcwCustomExt>() {
+			Some(ext) => {
+				let encoded = ext.0.clone().encode();
```
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply common/primitives/src/offchain.rs:39]: I kept the same logic as before, previously it was sent encoded but i think its just a matter of choice at this point

[the remark this answers is not in the document]
Comment (aramikm) [reply common/primitives/src/offchain.rs:39]: I think this is fine as is. I've tested it and it passes the correct values

File: common/primitives/src/offchain.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply common/primitives/src/offchain.rs:44]: @aramikm  haveing checked the I/O this works as ABI remains same for older function, i guess polkadot-sdk cant really ever remove AllocateAndReturnByCodec other than marking it as `Deprecated`


1. **ABI stays the same** → both before and after 2506 the host function compiles to

   ```
   () -> (i64)
   ```

   so old nodes/runtime calls won’t break.

2. **Rust type wrapper changed** → you can no longer return `Option<Vec<u8>>` directly,
   you must wrap it as `AllocateAndReturnByCodec<Option<Vec<u8>>>` to satisfy the new trait system.

[the remark this answers is not in the document]
Comment (aramikm) [reply common/primitives/src/offchain.rs:44]: That sounds great, a few things to consider. If that is the case then does it mean that the older runtime would also work with this kinda updated (with same signature) method!? If that is the case then I think we should return the correct value in get_val instead of None. This would help us in the case of New node and Old runtime. If that is the case we can still use the get_val in the `fetch_finalized_block_hash` and that should work as is. The idea is that we only are going to add `get_val_buffered` on the node side but we are not going to use it until the next runtime upgrade. That would make sure in no circumstance anything will break. But all of this is based on the idea that New Node (with `fn get_val(&mut self) -> AllocateAndReturnByCodec<Option<Vec<u8>>>` ) can work with old runtime as is without any issues.

[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply common/primitives/src/offchain.rs:44]: That is one test I will run, I will have to run two nodes and perform an upgrade keep one one, I will run that test next

**Comment by saraswatpuneet** [comment]
Awaiting resolution of [paritytech/polkadot-sdk#9498](https://github.com/paritytech/polkadot-sdk/issues/9498).
We are encountering edge cases in the fork-aware transaction pool, resulting in occasional timeouts—even when there are zero transactions in the pool.

File: .maintain/frame-system-extensions-weight-template.hbs [hunk 8]
```diff
@@ -1,4 +1,4 @@
-// Copied from: https://github.com/paritytech/polkadot-sdk/blob/polkadot-stable2503/substrate/.maintain/frame-weight-template.hbs
+// Copied from: https://github.com/paritytech/polkadot-sdk/blob/polkadot-stable2506-1/substrate/.maintain/frame-weight-template.hbs
```
Comment (mattheworris) [inline .maintain/frame-system-extensions-weight-template.hbs:1]: Note: There are some differences to this file in 2506-1, compared to our version, see [here](https://github.com/paritytech/polkadot-sdk/blob/polkadot-stable2506-1/substrate/.maintain/frame-weight-template.hbs).
Should we update this file?

**Review by mattheworris — APPROVED** [review]
- Read through changes.
- A couple of comments, good work on this chore.
🚢 it!

File: runtime/frequency/src/lib.rs [hunk 9]
```diff
@@ -2228,6 +2225,11 @@ sp_api::impl_runtime_apis! {
 				fn alias_origin() -> Result<(xcm::benchmarks::Location, xcm::benchmarks::Location), BenchmarkError> {
 					Err(BenchmarkError::Skip)
 				}
+
```
Comment (mattheworris) [inline runtime/frequency/src/lib.rs:2228]: nit: is this related to the sdk upgrade? Seems like an xcm related change.

**Review by mattheworris — APPROVED** [review]
- Read through changes, non-blocking comment.
🚢 it!

File: runtime/frequency/src/lib.rs (hunk 9 above)
Comment (saraswatpuneet) [reply runtime/frequency/src/lib.rs:2228]: @mattheworris yes these are newly added methods and following the change it seems this does not applies to us, @enddynayn  there are come changes here that relates to xcm

**Review by aramikm — APPROVED** [review]
Overall looks good. We need to make sure to upgrade all the nodes first and then we can do runtime upgrades due to using a new `hostfunction` inside the runtime.

File: .maintain/frame-system-extensions-weight-template.hbs [hunk 10]
```diff
@@ -42,43 +42,45 @@ use core::marker::PhantomData;
 
 /// Weights for `{{pallet}}` using the Substrate node and recommended hardware.
 pub struct WeightInfo<T>(PhantomData<T>);
-impl<T: frame_system::Config> frame_system::ExtensionsWeightInfo for WeightInfo<T> {
```
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply .maintain/frame-system-extensions-weight-template.hbs:45]: I think this is copied straight from polkadot, i will cross verify

[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply .maintain/frame-system-extensions-weight-template.hbs:45]: I copied the same as in pinned polkadot stable 2506-2 and ran the benchmarks again so looks good
